New package: dotnet-sdk-5.0.101

kawaiiamber opened the pull request:
# About
This addresses [this](https://github.com/void-linux/void-packages/pull/20147) and [this](https://github.com/void-linux/void-packages/issues/10773). I would have time to maintain this package. I would love to see it added to Void as I use dotnet everyday. I managed to get dotnet CORE on my musl system manually. I mainly just copied the template from the old PR linked, I may be doing stuff wrong. Cheers ~

ericonr commented on srcpkgs/dotnet-sdk/template:
```diff
@@ -0,0 +1,45 @@
+pkgname=dotnet-sdk
+version=5.0.101
+revision=1
+archs="x86_64 x84_64-musl"
```
```suggestion
archs="x86_64*"
```

kawaiiamber replied:
Done

kawaiiamber commented:
According to the checks:
```
=> dotnet-sdk-5.0.101_1: verifying checksum for distfile 'dotnet-sdk-5.0.101-linux-x64.tar.gz'... OK.
=> dotnet-sdk-5.0.101_1: running do-extract hook: 00-distfiles ...
=> dotnet-sdk-5.0.101_1: extracting distfile(s), please wait...
/void-packages/common/xbps-src/shutils/common.sh: line 31: cd: /builddir/dotnet-sdk-5.0.101_1: No such file or directory
=> ERROR: dotnet-sdk-5.0.101_1: cannot access wrksrc directory [/builddir/dotnet-sdk-5.0.101_1]
Error: Process completed with exit code 1.
```
I tried some different values of setting wrksrc, yet the same error occurs.

ev-ermakov commented:
> export CLR_OPENSSL_VERSION_OVERRIDE=48

This is a little sad to see.

A better alternative would be to build from source, however, this is a bit tricky...

kawaiiamber commented:
I think I see the issue. dotnet doesn't put stuff in it's own directory when downloading. I tried doing things like `wrksrc=${XBPS_BUILDDIR}`, but that just seems to literally put `/builddir` twice. I'm still a noob at xbps-src...

Anachron commented:
You can use create_wrksrc.

kawaiiamber commented:
`create_wrksrc=true`?

kawaiiamber commented:
It still seems to fail to access / create wrksrc after `create_wrksrc=yes`.

kawaiiamber commented:
```
=> dotnet-sdk-5.0.101_1: running post-install hook: 99-pkglint-warn-cross-cruft ...
=> dotnet-sdk-5.0.101_1: running pre-pkg hook: 03-rewrite-python-shebang ...
=> dotnet-sdk-5.0.101_1: running pre-pkg hook: 04-generate-runtime-deps ...
   SONAME: libstdc++.so.6 <-> libstdc++>=4.4.0_1
   SONAME: libgcc_s.so.1 <-> libgcc>=4.4.0_1
   SONAME: libc.musl-x86_64.so.1 <-> UNKNOWN PKG PLEASE FIX!
   SONAME: libmscordaccore.so <-> dotnet-sdk (ignored)
   SONAME: liblttng-ust.so.0 <-> lttng-ust>=2.7.2_1
   SONAME: libz.so.1 <-> zlib>=1.2.3_1
   SONAME: libgssapi_krb5.so.2 <-> mit-krb5-libs>=1.8_1
=> ERROR: dotnet-sdk-5.0.101_1: cannot guess required shlibs, aborting!
Error: Process completed with exit code 1.
```
Looking at the error, it seems like `SONAME: libc.musl-x86_64.so.1 <-> UNKNOWN PKG PLEASE FIX!` is what I need to fix. I wonder why it's unknown?

kawaiiamber commented:
I could just be doing this all wrong, but, I'm really curious why it's reporting `SONAME: libc.musl-x86_64.so.1 <-> UNKNOWN PKG PLEASE FIX!` on the musl arch. I copied the info from `/etc/dotnet.sh` to my `$HOME/.zshrc` and, after exporting the `$PATH`, everything seems to work fine on my end. I ran `ldd /etc/dotnet/dotnet` and, on my musl machine, it reports
```
[amber@otaku ~]$ ldd /opt/dotnet/dotnet
        /lib/ld-musl-x86_64.so.1 (0x7fb35c13f000)
        libstdc++.so.6 => /lib/libstdc++.so.6 (0x7fb35be94000)
        libgcc_s.so.1 => /lib/libgcc_s.so.1 (0x7fb35be7b000)
        libc.musl-x86_64.so.1 => /lib/ld-musl-x86_64.so.1 (0x7fb35c13f000)
```
It looks like `libc.musl-x86_64.so.1` is successfully being linked to `/lib/ld-musl-x86_64.so.1` on my end, yet is failing durring the checks. I'd love to hear others thoughts ~

ericonr commented:
Somehow they encoded into the binary that libc is `libc.musl-x86-64.so`... If you can launch the application on your system, you could instead just add `skiprdeps="opt/dotnet/dotnet"` to the template.

kawaiiamber commented:
Thank you for the suggestion, I really appreciate how friendly and helpful you are. I added `skiprdeps="opt/dotnet/dotnet"` to the template.

kawaiiamber commented:
Perhaps I put it in incorrectly, but the log is still reporting
```
=> dotnet-sdk-5.0.101_1: running post-install hook: 13-pkg-config-clean-xbps-cross-base-ref ...
=> dotnet-sdk-5.0.101_1: running post-install hook: 99-pkglint-warn-cross-cruft ...
=> dotnet-sdk-5.0.101_1: running pre-pkg hook: 03-rewrite-python-shebang ...
=> dotnet-sdk-5.0.101_1: running pre-pkg hook: 04-generate-runtime-deps ...
   SONAME: libstdc++.so.6 <-> libstdc++>=4.4.0_1
   SONAME: libgcc_s.so.1 <-> libgcc>=4.4.0_1
   SONAME: libc.musl-x86_64.so.1 <-> UNKNOWN PKG PLEASE FIX!
   SONAME: libz.so.1 <-> zlib>=1.2.3_1
   SONAME: libgssapi_krb5.so.2 <-> mit-krb5-libs>=1.8_1
   SONAME: libmscordaccore.so <-> dotnet-sdk (ignored)
   SONAME: liblttng-ust.so.0 <-> lttng-ust>=2.7.2_1
=> ERROR: dotnet-sdk-5.0.101_1: cannot guess required shlibs, aborting!
Error: Process completed with exit code 1.
```
I'm sorry if I'm doing this wrong / wasting your time.

kawaiiamber commented:
I added `skiprdeps="${DESTDIR}/opt/dotnet/dotnet"` to the template. Now it's successfully saying `=> Skipping dependency scan for /opt/dotnet/dotnet`, but it looks like it's still failing to link musl libc?
```
[amber@otaku ~/projx/repos/void-packages]$ ./xbps-src pkg dotnet-sdk
=> xbps-src: updating repositories for host (x86_64-musl)...
[*] Updating repository `https://alpha.de.repo.voidlinux.org/current/musl/x86_64-musl-repodata' ...
[*] Updating repository `https://alpha.de.repo.voidlinux.org/current/musl/nonfree/x86_64-musl-repodata' ...
[*] Updating repository `https://alpha.de.repo.voidlinux.org/current/musl/debug/x86_64-musl-repodata' ...
=> xbps-src: updating software in / masterdir...
=> xbps-src: cleaning up / masterdir...
=> dotnet-sdk-5.0.101_1: removing autodeps, please wait...
=> dotnet-sdk-5.0.101_1: building for x86_64-musl...
=> dotnet-sdk-5.0.101_1: skipping check (XBPS_CHECK_PKGS is disabled) ...
=> dotnet-sdk-5.0.101_1: running pre-pkg hook: 03-rewrite-python-shebang ...
=> dotnet-sdk-5.0.101_1: running pre-pkg hook: 04-generate-runtime-deps ...
=> Skipping dependency scan for /opt/dotnet/dotnet
   SONAME: libstdc++.so.6 <-> libstdc++>=4.4.0_1
   SONAME: libgcc_s.so.1 <-> libgcc>=4.4.0_1
   SONAME: libc.musl-x86_64.so.1 <-> UNKNOWN PKG PLEASE FIX!
   SONAME: libz.so.1 <-> zlib>=1.2.3_1
   SONAME: libgssapi_krb5.so.2 <-> mit-krb5-libs>=1.8_1
   SONAME: libmscordaccore.so <-> dotnet-sdk (ignored)
   SONAME: liblttng-ust.so.0 <-> lttng-ust>=2.7.2_1
=> ERROR: dotnet-sdk-5.0.101_1: cannot guess required shlibs, aborting!
```

idMysteries commented:
Try this
ln -s /usr/lib/x86_64-linux-musl/libc.so /lib/libc.musl-x86_64.so.1

ericonr commented:
Hm, maybe it should be `skiprdeps=/opt...` (I missed the `/` in the beginning).

That said, might be easier to just put `shlib_provides="libc.musl-x86_64.so.1"` in the template for musl.

kawaiiamber commented:
I added `shlib_provides="libc.musl-x86_64.so.1"` in the case for musl, yet the same result.

unspecd commented:
Try this:
```sh
	# The binary somehow has bad link, so this seems to be needed to fix it
        skiprdeps="
	  ${_target}/dotnet
	  ${_target}/host/fxr/5.0.1/libhostfxr.so
          ${_target}/packs/Microsoft.NETCore.App.Host.linux-musl-x64/5.0.1/runtimes/linux-musl-x64/native/apphost
          ${_target}/packs/Microsoft.NETCore.App.Host.linux-musl-x64/5.0.1/runtimes/linux-musl-x64/native/singlefilehost
	  ${_target}/packs/Microsoft.NETCore.App.Host.linux-musl-x64/5.0.1/runtimes/linux-musl-x64/native/libnethost.so
	  ${_target}/sdk/5.0.101/AppHostTemplate/apphost
	  ${_target}/shared/Microsoft.NETCore.App/5.0.1/createdump
	  ${_target}/shared/Microsoft.NETCore.App/5.0.1/libSystem.IO.Compression.Native.so
	  ${_target}/shared/Microsoft.NETCore.App/5.0.1/libSystem.Native.so
	  ${_target}/shared/Microsoft.NETCore.App/5.0.1/libSystem.Net.Security.Native.so
	  ${_target}/shared/Microsoft.NETCore.App/5.0.1/libSystem.Security.Cryptography.Native.OpenSsl.so
	  ${_target}/shared/Microsoft.NETCore.App/5.0.1/libclrjit.so
	  ${_target}/shared/Microsoft.NETCore.App/5.0.1/libcoreclr.so
	  ${_target}/shared/Microsoft.NETCore.App/5.0.1/libcoreclrtraceptprovider.so
	  ${_target}/shared/Microsoft.NETCore.App/5.0.1/libdbgshim.so
	  ${_target}/shared/Microsoft.NETCore.App/5.0.1/libhostpolicy.so
	  ${_target}/shared/Microsoft.NETCore.App/5.0.1/libmscordaccore.so
	  ${_target}/shared/Microsoft.NETCore.App/5.0.1/libmscordbi.so
	"
```

kawaiiamber commented:
Holy wow! Thank you so much! I feel bad that you did all of that to fix my problem, thank you x1000 ~

unspecd commented:
Glad to help you out :)

fosslinux commented:
This is a binary package, no? It should have the suffix `-bin`.

fosslinux commented on srcpkgs/dotnet-sdk/template:
```diff
@@ -0,0 +1,75 @@
+# Template file for 'dotnet-sdk'
+pkgname=dotnet-sdk
```
As per my other comment I think this should be `dotnet-sdk-bin`.

kawaiiamber replied:
That's a good point. Renamed to `dotnet-sdk-bin`

fosslinux commented on srcpkgs/dotnet-sdk/template:
```diff
@@ -0,0 +1,75 @@
+pkgname=dotnet-sdk
+version=5.0.101
+revision=1
+archs="x86_64 aarch64 x86_64-musl"
```
`x86_64*`

kawaiiamber replied:
I thought I already did that... I must have accidently reverted back at some point. Thanks for pointing that out. Updated ~

fosslinux commented on srcpkgs/dotnet-sdk/template:
```diff
@@ -0,0 +1,75 @@
+license="MIT"
+homepage="https://dotnet.microsoft.com"
+
+_target=${DESTDIR}/opt/dotnet
```
Hm, do we really want it in `/opt/dotnet`? I don't think we generally have packages in `/opt`. How about `/usr/libexec/dotnet`?

kawaiiamber replied:
While packages generally don't go in `/opt`, given all the various things that the dotnet folder contains and does, I don't know if `/usr/libexec` is a better place. Though, I'm not very experienced and am probably wrong. I'd love to hear your thoughts on this. Thanks ~

q66 replied:
keeping `/opt` is fine given this is going to be a restricted package from binaries

kawaiiamber replied:
Sounds good

fosslinux commented on srcpkgs/dotnet-sdk/template:
```diff
@@ -0,0 +1,75 @@
+
+do_install() {
+	vmkdir ${_target}
+	vcopy ${wrksrc}/packs ${_target}
```
Shouldn't need `${wrksrc}/` here, I think

kawaiiamber replied:
Removed redundant `${wrksrc}` expansion.

kawaiiamber commented:
Changed `archs="x86_64 aarch64 x86_64-musl"` to `archs="x86_64* aarch64"` and removed redundant `${wrksrc}` expansion. The way that dotnet is made is a little awkward. I'm still convinced `/opt` is probably the best place for it given the various things that the files do. I could be wrong though. I don't see how '/usr/libexec' is better than `/opt` in this case, though I'm probably wrong. An explanation would be greatly appreciated. Cheers ~

q66 commented:
pointing to the libc thing to ld-musl looks pretty wrong, it shouldn't really, ignoring rdeps won't magically make it work or be right

it needs to be investigated why that name is there (I don't think it exists anywhere in any form, unless it's an internal link in dotnet, in which case it's probably okay)

kawaiiamber commented:
@q66, without `skiprdeps` on the `musl` arch, `xbps-src` and the checks here report:
```sh
=> dotnet-sdk-bin-5.0.101_1: running post-install hook: 10-pkglint-devel-paths ...
=> dotnet-sdk-bin-5.0.101_1: running post-install hook: 11-pkglint-elf-in-usrshare ...
=> dotnet-sdk-bin-5.0.101_1: running post-install hook: 12-rename-python3-c-bindings ...
=> dotnet-sdk-bin-5.0.101_1: running post-install hook: 13-pkg-config-clean-xbps-cross-base-ref ...
=> dotnet-sdk-bin-5.0.101_1: running post-install hook: 99-pkglint-warn-cross-cruft ...
=> dotnet-sdk-bin-5.0.101_1: running pre-pkg hook: 03-rewrite-python-shebang ...
=> dotnet-sdk-bin-5.0.101_1: running pre-pkg hook: 04-generate-runtime-deps ...
   SONAME: libstdc++.so.6 <-> libstdc++>=4.4.0_1
   SONAME: libgcc_s.so.1 <-> libgcc>=4.4.0_1
   SONAME: libc.musl-x86_64.so.1 <-> UNKNOWN PKG PLEASE FIX!
   SONAME: libz.so.1 <-> zlib>=1.2.3_1
   SONAME: libgssapi_krb5.so.2 <-> mit-krb5-libs>=1.8_1
   SONAME: libmscordaccore.so <-> dotnet-sdk-bin (ignored)
   SONAME: liblttng-ust.so.0 <-> lttng-ust>=2.7.2_1
=> ERROR: dotnet-sdk-bin-5.0.101_1: cannot guess required shlibs, aborting!
```
However, on my musl machine, the binaries successfully run and link:
```sh
[amber@otaku ~]$ ldd /opt/dotnet/dotnet
        /lib/ld-musl-x86_64.so.1 (0x7ff2b8419000)
        libstdc++.so.6 => /lib/libstdc++.so.6 (0x7ff2b816e000)
        libgcc_s.so.1 => /lib/libgcc_s.so.1 (0x7ff2b8155000)
        libc.musl-x86_64.so.1 => /lib/ld-musl-x86_64.so.1 (0x7ff2b8419000)
```
Thus, I suspect that it's an internal link in dotnet us you point out.

q66 commented:
the output is a red herring, there is no relation between the two names and dynlinker should never be in the dependencies either

but idk what is going on there

kawaiiamber commented:
I'm not too sure myself. I'm following some suggestions from previous comments in this thread that seem to fix it for the `musl` arch.

q66 commented:
yeah i'm not a fan of having to list them all out, it's verbose and adds lots of clutter, maybe it'd be worth it to find out what is really going on

also maybe one could do like `shlib_provides="libc.musl-x86_64.so.1"` as a workaround instead...

kawaiiamber commented:
Removing `skiprdeps` and replacing it with `shlib_provides="libc.musl-x86_64.so.1"` results in the following from `xbps-src`:
```sh
=> dotnet-sdk-bin-5.0.101_1: running post-install hook: 10-pkglint-devel-paths ...
=> dotnet-sdk-bin-5.0.101_1: running post-install hook: 11-pkglint-elf-in-usrshare ...
=> dotnet-sdk-bin-5.0.101_1: running post-install hook: 12-rename-python3-c-bindings ...
=> dotnet-sdk-bin-5.0.101_1: running post-install hook: 13-pkg-config-clean-xbps-cross-base-ref ...
=> dotnet-sdk-bin-5.0.101_1: running post-install hook: 99-pkglint-warn-cross-cruft ...
=> dotnet-sdk-bin-5.0.101_1: running pre-pkg hook: 03-rewrite-python-shebang ...
=> dotnet-sdk-bin-5.0.101_1: running pre-pkg hook: 04-generate-runtime-deps ...
   SONAME: libstdc++.so.6 <-> libstdc++>=4.4.0_1
   SONAME: libgcc_s.so.1 <-> libgcc>=4.4.0_1
   SONAME: libc.musl-x86_64.so.1 <-> UNKNOWN PKG PLEASE FIX!
   SONAME: libz.so.1 <-> zlib>=1.2.3_1
   SONAME: libgssapi_krb5.so.2 <-> mit-krb5-libs>=1.8_1
   SONAME: libmscordaccore.so <-> dotnet-sdk-bin (ignored)
   SONAME: liblttng-ust.so.0 <-> lttng-ust>=2.7.2_1
=> ERROR: dotnet-sdk-bin-5.0.101_1: cannot guess required shlibs, aborting!
```

q66 commented:
yeah you'd also need to add it in `common/shlibs` (see how an equivalent is done for e.g. electron-using programs, e.g. `atom` and `libGLESv2.so`

kawaiiamber commented:
I see. It seems that there are two entries per line in common/shlibs. I assume the line I would add would looks like:
```
libc.musl-x86_64.so.1 ld-musl-x86_64.so.1
```

q66 commented:
the second entry is the package that provides it

so it would be like,

`libc.musl-x86_64.so.1 dotnet-sdk-bin-5.0.101_1`

the file is used to map shlib names back to package names

kawaiiamber commented:
Oop, I appear to have done something wrong when adding `libc.musl-x86_64.so.1 dotnet-sdk-bin-5.0.101_1` to common/shlibs. It looks as though there's a conflict. It seems to pass locally with `./xbps-src pkg dotnet-sdk-bin` though.

kawaiiamber commented:
I wonder why there's a conflict on github. It reports that
>This branch has conflicts that must be resolved
>Only those with write access to this repository can merge pull requests.
>Conflicting files
>common/shlibs

I added the line to common/shlibs normally. Is there something I'm not seeing here?

q66 commented:
probably conflicts with current master branch? that's why when modifying the file you generally add your changes somewhere in the middle and not at the end

kawaiiamber commented:
Moved `libc.musl-x86_64.so.1 dotnet-sdk-bin-5.0.101_1` higher up the common/shlib file

ericonr commented on srcpkgs/dotnet-sdk-bin/template:
```diff
@@ -0,0 +1,56 @@
+license="MIT"
+homepage="https://dotnet.microsoft.com"
+
+_target=${DESTDIR}/opt/dotnet
```
Please put this line together with `_prof`, down below.

kawaiiamber replied:
Done

kawaiiamber commented on srcpkgs/dotnet-sdk-bin/template:
```diff
@@ -0,0 +1,56 @@
+	checksum=1ecf97cab3d9d38af32ee0cbaaa41315a6896dc0980218ac227329f989a5ef51
+	# The binary somehow has bad link in the binary, so this is added to common/shlibs
+	shlib_provides="libc.musl-x86_64.so.1"
+	;;
```
Done

kawaiiamber commented on srcpkgs/dotnet-sdk-bin/template:
```diff
@@ -0,0 +1,56 @@
+distfiles="https://download.visualstudio.microsoft.com/download/pr/${_path}/dotnet-sdk-${version}-linux-${_arch}.tar.gz"
+
+_prof=${DESTDIR}/etc/profile.d
+wrksrc=dotnet-sdk-${version}-linux-${_arch}
```
Done

kawaiiamber commented on srcpkgs/dotnet-sdk-bin/files/dotnet.sh:
```diff
@@ -0,0 +1,6 @@
+export DOTNET_ROOT=/opt/dotnet
+export PATH="$PATH:$DOTNET_ROOT"
+# This seems to be needed to be able to work with LibreSSL
+export CLR_OPENSSL_VERSION_OVERRIDE=48
```
Indeed, it would seem as if the LibreSSL updates might break this override is the version isn't properly maintainted. I can't think of a better work around for the time being. I'd love to hear your thoughts ~

ericonr replied:
I'm not sure there is a proper non hacky one :/

kawaiiamber replied:
That does seem to be an issue working with these binaries over building from source. Given that this is the best one probably can do with these pre-built binaries, is it okay to resolve the converstaion?

kawaiiamber commented:
Thank you so much for the review and suggestions - they were all very helpful! :)

q66 commented:
so, is there any reason not to build this from source? since it seems to be MIT licensed, i don't see one, and void generally doesn't ship third party binaries of things that are not proprietary

q66 commented:
we also found out the source of the curious dependency - it seems alpine ships this in their musl

kawaiiamber commented:
I was curious about building the framework from source myself. The source is [here](https://github.com/dotnet/core) with the release for the source version `5.0.1` [here](https://github.com/dotnet/core/releases/tag/v5.0.1). This seems to be much more trickly and difficult, however. I would love to hear your thoughts ~

ericonr commented:
The issue, to my knowledge, is lack of LibreSSL support. That it works at all is pretty interesting by itself.

kawaiiamber commented:
@ericonr, agreed. An ideal solution would be to build the framework from source, but this seems to be remarkably difficult to do. If one was able to build it from source, it would then be easier to implement LibreSSL support directly. While this solution of overriding the SSL used by dotnet binaries isn't ideal, it's the best I can think of. Thanks for sharing your thoughts ~

q66 commented:
it doesn't necessarily mean it actually works, it just means that it works enough that you haven't run into any ABI issues yet - but this is pretty dangerous, ABI mismatch can result in some pretty nasty and often subtle issues

kawaiiamber commented:
I see. While this is completely anecdotal, I have yet to see bugs on my musl or glibc systems using this solution. The risk of an ABI issue is pretty significant, however. If it really risks ABI issues, I wonder if the dotnet sdk binaries should even be packaged in the first place, then. To be safe, it seems like the best solution is to build the framework from source. I have no experience at ABI kinds of things - I don't know the scope / gravity when it comes to the risks of implimenting the framework this way. Thanks for the thoughts ~

q66 commented:
either way with concerns like this i'm reluctant to allow this ino the repository; we could possibly do `restricted=yes` on this, and maybe figure out building from source later, maybe after the move back to openssl is done

kawaiiamber commented:
Should I go ahead and add `restricted=yes` to the template?

q66 commented:
i'd go ahead and do that, plus a comment explaining the reason

ABI issues can be subtle and can result in e.g. junk data being passed where it shouldn't, and especially with crypto/tls libraries you definitely don't want security issues arising from this

kawaiiamber commented:
Understood - added `resctricted=yes` to the template file with a comment explaning.

ericonr commented:
Now that we are definitely moving to OpenSSL and NET folks are in talks with Alpine to get it officially built there, I'd consider waiting a bit to see if we can use their work to make our own build from source.

kawaiiamber commented:
> Now that we are definitely moving to OpenSSL and NET folks are in talks with Alpine to get it officially built there, I'd consider waiting a bit to see if we can use their work to make our own build from source.

I tried messing around with [this](https://github.com/dotnet/source-build) repo to see if I can't get it to build on VOID. I'm curious to see wha the Alpine build will look like.

fosslinux commented:
I would be interested in attempting to get it to build from source once OpenSSL is merged.

kawaiiamber commented:
> I would be interested in attempting to get it to build from source once OpenSSL is merged.

Now that OpenSSL is merged over LibreSSL, work can probably start on building from source. I'm curious about [this](https://github.com/dotnet/source-build) repo from building from source. There is no release tag, however.

Anachron commented:
What do you mean? It has [release tags](https://github.com/dotnet/source-build/tags).
Interesting repo, I've star-ed it and I'll see how well the build-system there works. From a quick glance it's not really different than the other offical repo.

kawaiiamber commented:
> What do you mean? It has [release tags](https://github.com/dotnet/source-build/tags).
> Interesting repo, I've star-ed it and I'll see how well the build-system there works. From a quick glance it's not really different than the other offical repo.

Oops, I missed those. I must admit - I'm not too familiar with how github works. I didn't see the tags on the right and assumed they weren't there.

kawaiiamber commented:
Even though I'm using a release, the ./build.sh script still seems to spit an error when git is missing, but adding git to hostmakedepends, it seems to error saying it's not a git repository? Scratching my head on this one.

ericonr commented:
You seem to be missing `makedepends`, which would include at least `icu-devel`.

kawaiiamber commented:
> You seem to be missing `makedepends`, which would include at least `icu-devel`.

Added `icu-devel` to `makedepends`. Also, possibly noob question, but - is there a way to not have it push through CI everytime since there's not a working build, yet?

ericonr commented:
Add `[ci skip]` to the PR description (not the title).

kawaiiamber commented:
> Add `[ci skip]` to the PR description (not the title).

Okay, added `[ci skip]`

rmboggs commented:
@kawaiiamber , thank you for taking the time to work on this. It's one of the few remaining items that i need in order to use void as my primary system.

I did some testing on this using an x64 glib system i setup this week.  I was able to get farther in the build by adding more needed make depends that the build was bombing on. However, im now stuck on an undefined symbol error on the zlib library.  Ill post the make depends and current error msg soon, if it will help.

kawaiiamber commented:
> @kawaiiamber , thank you for taking the time to work on this. It's one of the few remaining items that i need in order to use void as my primary system.
> 
> I did some testing on this using an x64 glib system i setup this week. I was able to get farther in the build by adding more needed make depends that the build was bombing on. However, im now stuck on an undefined symbol error on the zlib library. Ill post the make depends and current error msg soon, if it will help.

Ya - any help would be nice haha. Biggest thing at the moment is I'm not sure how to handle `RUNTIME_ID` on void - since it seems to be hard coded into the build script for dotnet? As for undefined symbols on zlib libraries, did you add `zlib-devel` package to `makedepends`? That would be my guess.

rmboggs commented:
> > @kawaiiamber , thank you for taking the time to work on this. It's one of the few remaining items that i need in order to use void as my primary system.
> > 
> > I did some testing on this using an x64 glib system i setup this week. I was able to get farther in the build by adding more needed make depends that the build was bombing on. However, im now stuck on an undefined symbol error on the zlib library. Ill post the make depends and current error msg soon, if it will help.
> 
> Ya - any help would be nice haha. Biggest thing at the moment is I'm not sure how to handle `RUNTIME_ID` on void - since it seems to be hard coded into the build script for dotnet? As for undefined symbols on zlib libraries, did you add `zlib-devel` package to `makedepends`? That would be my guess.

I did, that was the last package i added to make depends when this error came up.  I'm wondering if it is defined in a different obscure package or something.

As for the RUNTIME_ID, you can't force it to use linux-x64/linux-musl-x64?

rmboggs commented:
This is what I currently have in the hostmakedepends and makedepends lines of your template file.  Again, this worked for me on my x86_64 glibc system, so I am not 100% sure if this will work on the other platforms:
```bash
hostmakedepends="git tar wget python3 cmake clang llvm"
makedepends="icu-devel libunwind-devel lttng-ust-devel zlib-devel mit-krb5-devel"
```
I'm still new to the void package process so if I have these items listed incorrectly, please let me know so I can correct and learn. :)

Regarding the error, it's a bit weird:
```
/usr/bin/x86_64-unknown-linux-gnu-ld: /builddir/source-build-5.0.202-SDK/artifacts/src/runtime.2f740adc1457e8a28c1c072993b66f515977eb51/artifacts/bin/native/net5.0-Linux-Release-x64/libSystem.IO.Compression.Native.a(pal_zlib.c.o): in function 'CompressionNative_DeflateInit2_':
/builddir/source-build-5.0.202-SDK/artifacts/src/runtime.2f740adc1457e8a28c1c072993b66f515977eb51/artifacts/bin/native/AnyOS/zlib/pal_zlib.c:120: undefined reference to 'deflateInit2_'
```
It's weird because I do see this defined in the zlib.h file when I install zlib on my system.  I must be doing something wrong here.  Any thoughts?

Anachron commented:
Can confirm, builds continues here with the dependencies from https://github.com/void-linux/void-packages/pull/27170#issuecomment-839754296.

But fails for me with:
```
    /usr/bin/x86_64-unknown-linux-gnu-ld: /builddir/source-build-5.0.202-SDK/artifacts/src/runtime.2f740adc1457e8a28c1c072993b66f515977eb51/src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c:565: undefined reference to `gss_release_oid_set'
    clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
    make[2]: *** [cli/apphost/static/CMakeFiles/singlefilehost.dir/build.make:287: cli/apphost/static/singlefilehost] Error 1
    make[1]: *** [CMakeFiles/Makefile2:439: cli/apphost/static/CMakeFiles/singlefilehost.dir/all] Error 2
```

One thing though that I've recognized is that dotnet seems to work with one CORE only? clang++ does not have this "issue".

TinfoilSubmarine commented:
I saw that the Arch package had a patch for the linker errors with zlib/gss, so I created a similar one. This patch is used by the `source-build` scripts, not by Void, and needs to be placed in `patches/runtime/` in the build wrksrc. For simplicity instead of trying to verbally explain all my changes, [I pushed all these changes to my fork here](https://github.com/TinfoilSubmarine/void-packages/commit/cce7e1fd99e4015a6c79ea85153b28472ce4e9ee).

Still getting another error afterwards though:
```
  /builddir/source-build-5.0.206.1-SDK/.dotnet/sdk/5.0.205/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.targets(257,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Runtime.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/builddir/source-build-5.0.206.1-SDK/artifacts/src/aspnetcore.c663adee8e64ba5d379fa0edfb8201984a7df7d0/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
  /builddir/source-build-5.0.206.1-SDK/.dotnet/sdk/5.0.205/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.targets(257,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Runtime.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/builddir/source-build-5.0.206.1-SDK/artifacts/src/aspnetcore.c663adee8e64ba5d379fa0edfb8201984a7df7d0/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
  /builddir/source-build-5.0.206.1-SDK/.dotnet/sdk/5.0.205/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.targets(257,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Runtime.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/builddir/source-build-5.0.206.1-SDK/artifacts/src/aspnetcore.c663adee8e64ba5d379fa0edfb8201984a7df7d0/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
  /builddir/source-build-5.0.206.1-SDK/.dotnet/sdk/5.0.205/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.targets(257,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Runtime.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/builddir/source-build-5.0.206.1-SDK/artifacts/src/aspnetcore.c663adee8e64ba5d379fa0edfb8201984a7df7d0/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
  /builddir/source-build-5.0.206.1-SDK/.dotnet/sdk/5.0.205/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.targets(257,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Runtime.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/builddir/source-build-5.0.206.1-SDK/artifacts/src/aspnetcore.c663adee8e64ba5d379fa0edfb8201984a7df7d0/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
  /builddir/source-build-5.0.206.1-SDK/.dotnet/sdk/5.0.205/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.targets(257,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Runtime.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/builddir/source-build-5.0.206.1-SDK/artifacts/src/aspnetcore.c663adee8e64ba5d379fa0edfb8201984a7df7d0/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
  /builddir/source-build-5.0.206.1-SDK/artifacts/src/aspnetcore.c663adee8e64ba5d379fa0edfb8201984a7df7d0/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj(401,5): error : Could not find crossgen tools/crossgen
      0 Warning(s)
      7 Error(s)
  
  Time Elapsed 00:01:46.85
  ##vso[task.logissue type=error]Build failed (exit code '1').
  
  'aspnetcore' failed during build.
  See '/builddir/source-build-5.0.206.1-SDK/artifacts/logs/aspnetcore.log' for more information.
  'known-good' failed during build.

Build FAILED.

/builddir/source-build-5.0.206.1-SDK/packages/restored/microsoft.dotnet.arcade.sdk/5.0.0-beta.20426.4/tools/ProjectLayout.props(8,3): warning MSB4011: "/builddir/source-build-5.0.206.1-SDK/packages/restored/microsoft.dotnet.arcade.sdk/5.0.0-beta.20426.4/tools/RepoLayout.props" cannot be imported again. It was already imported at "/builddir/source-build-5.0.206.1-SDK/packages/restored/microsoft.dotnet.arcade.sdk/5.0.0-beta.20426.4/tools/Build.proj (49,3)". This is most likely a build authoring error. This subsequent import will be ignored. 
/builddir/source-build-5.0.206.1-SDK/repos/Directory.Build.targets(374,5): error MSB3073: The command "/builddir/source-build-5.0.206.1-SDK/artifacts/src/aspnetcore.c663adee8e64ba5d379fa0edfb8201984a7df7d0/build.sh  --restore --build --pack --configuration Release --ci -bl /v:minimal --arch x64 /p:BuildNodeJs=false /p:SourceBuildRuntimeIdentifier=linux-x64 /p:UseAppHost=false /p:PublishCompressedFilesPathPrefix=/builddir/source-build-5.0.206.1-SDK/artifacts/obj/x64/Release/blobs/aspnetcore/Runtime/ /p:MicrosoftNetFrameworkReferenceAssembliesVersion=1.0.0  /p:DotNetPackageVersionPropsPath=/builddir/source-build-5.0.206.1-SDK/artifacts/obj/x64/Release/PackageVersions.props >> /builddir/source-build-5.0.206.1-SDK/artifacts/logs/aspnetcore.log 2>&1" exited with code 1. [/builddir/source-build-5.0.206.1-SDK/repos/aspnetcore.proj]
    1 Warning(s)
    1 Error(s)

Time Elapsed 00:02:13.18
Build failed (exit code '1').
```

kawaiiamber commented:
Sounds neat ^-^, I'll take a look when I get home.

rmboggs commented:
Ill check them out too tomorrow AM.  I just put Void back on my laptop this morning but i still need to setup my build environment.

TinfoilSubmarine commented:
I played around a little more and came across a patch in `patches/aspnetcore` that is seemingly problematic and removing it makes the build succeed for my machine. [See latest commit on my branch.](https://github.com/TinfoilSubmarine/void-packages/commit/6aafae2af427b9bee445abf4f0873c7478a2e273)

TinfoilSubmarine commented:
Also, I'm thinking about packaging and was comparing with the Arch PKGBUILD. Unless we want to split out into the individual components like Arch (`dotnet-host`, `dotnet-sdk`, `dotnet-runtime`, etc...), this package should probably be renamed to `dotnet` or something.

rmboggs commented:
Hi @TinfoilSubmarine,

I'm tried out your work on this yesterday and it bombed but after looking at the error message, i think the issue was related to internet connectivity (my internet has been spotty lately) so I am giving it another try this morning. I think it got farther than before so that is a good sign. 

As far as packaging, i am not sure what is best here. That's probably a question that should discussed further, at least when this current package can finally build on Void.

TinfoilSubmarine commented:
[Latest commit](https://github.com/TinfoilSubmarine/void-packages/commit/331ce71676460500d040aaa41965a19cd308666a) actually results in a successfully built package, all housed in `/usr/lib/dotnet`.

rmboggs commented:
> [Latest commit](https://github.com/TinfoilSubmarine/void-packages/commit/331ce71676460500d040aaa41965a19cd308666a) actually results in a successfully built package, all housed in `/usr/lib/dotnet`.

I'll try this too but it just bombed again during the post build step, which is alot farther than I got previously so that's awesome.  Error I got is basically saying that it couldn't find the artifact tar.gz file.  Looks like it's because the tar.gz file that it's looking for has a slightly different name than it was expecting due to the 5.0.206 **.1** in the version variable.  I'll see if I can tweak the template a bit to exclude the revision part of the version during the post build.  I'll let you know.

rmboggs commented:
Oh, looks like I am working off an older revision of the template file.  My mistake.  I'll try again.

rmboggs commented:
That was it.  After I grabbed the latest template file from @TinfoilSubmarine 's [dotnet-sdk repo](https://github.com/TinfoilSubmarine/void-packages/tree/dotnet-sdk), it built and installed successfully. YAY!  I'm going to try using it for my normal day-to-day workload and see how it goes.  

One thing I did notice was that it installed in /usr/lib/dotnet but there was nothing put in /usr/bin.  I get that but could a link to /usr/lib/dotnet/dotnet be put in /usr/bin?

TinfoilSubmarine commented:
I think the symlink would be ideal since then it may not need the `profile.d/dotnet.sh` path hack.

rmboggs commented:
> I think the symlink would be ideal since then it may not need the `profile.d/dotnet.sh` path hack.

I agree.  Is that something that can be added to the template script?

Anachron commented:
Good work, guys! I'll try building it again this weekend (hopefully).

TinfoilSubmarine commented:
Ok, [latest commit](https://github.com/TinfoilSubmarine/void-packages/commit/a9687054f2a88031392c80e8cef876083ae265a7) removes the `profile.d/dotnet.sh` and symlinks the executable into `/usr/bin`.

rmboggs commented:
> Ok, [latest commit](https://github.com/TinfoilSubmarine/void-packages/commit/a9687054f2a88031392c80e8cef876083ae265a7) removes the `profile.d/dotnet.sh` and symlinks the executable into `/usr/bin`.

This is great, thanks so much @TinfoilSubmarine.  I don't think this would have gotten this far without all your contributions here.  Much appreciated.

I have been able to build one of my projects with this and it seems to be working well with VS Code. :+1: 

Should we discuss either renaming the package and/or splitting the package up now or just worry about importing @TinfoilSubmarine's changes to the repo that this request is pulling from?

TinfoilSubmarine commented:
I don't have a strong opinion on whether this should be split or not. Doing it would be straightforward using subpackages, but I'm honestly not that familiar with the different components of dotnet and how they are used, so my (ignorant) preference would be to keep it simple and in one package.

The obvious argument for splitting is package size: currently the package file is 103MB, and 335MB installed. Breakdown on disk looks like this:
```
$ du -h --max-depth=1 /usr/lib/dotnet
4.3M	/usr/lib/dotnet/templates
85M	/usr/lib/dotnet/shared
173M	/usr/lib/dotnet/sdk
73M	/usr/lib/dotnet/packs
336K	/usr/lib/dotnet/host
335M	/usr/lib/dotnet
```

TinfoilSubmarine commented:
I should also mention that this is only working on x86_64* currently. I haven't been able to get cross-builds working using `./build.sh /p:Platform="${_ARCHITECTURE}"` since it tries to use "qemu-debootstrap" which I haven't seen before. [As such I removed the cross-build relevant lines and added `archs="x86_64*"`.](https://github.com/void-linux/void-packages/commit/bb9b4dded60def86853ce9d595c87695a120587f)

rmboggs commented:
> I should also mention that this is only working on x86_64* currently. I haven't been able to get cross-builds working using `./build.sh /p:Platform="${_ARCHITECTURE}"` since it tries to use "qemu-debootstrap" which I haven't seen before. [As such I removed the cross-build relevant lines and added `archs="x86_64*"`.](https://github.com/void-linux/void-packages/commit/bb9b4dded60def86853ce9d595c87695a120587f)

Yeah, I was going to mention that.  I have no way of checking the arm64 build either.  Probably good to only include x86_64 for the time being until someone can come in and get the arm64 build working at a later time.

rmboggs commented:
In regard to keeping the package simple, do you think we should just remove the "-sdk" from the package name then?

kawaiiamber commented:
> Also, I'm thinking about packaging and was comparing with the Arch PKGBUILD. Unless we want to split out into the individual components like Arch (`dotnet-host`, `dotnet-sdk`, `dotnet-runtime`, etc...), this package should probably be renamed to `dotnet` or something.

That's the plan - for this pacakge to remain `dotnet-sdk`.

TinfoilSubmarine commented:
Ok, that's fine if you want this to be just the sdk. 

Not sure if you're aware, but this build script also builds the other components, so if we want to have an individual package for the sdk, it would need to be a subpackage, along with subpackages for aspnet-runtime, dotnet-runtime, etc. All of them are included in the tarball generated by this script.

rmboggs commented:
Well, i was poking around in the repo that this new package pulls the source from and found another tag just for the dotnet runtime. If I am understanding this correctly, then it would make sense to just have 2 dotnet packages, one for sdk and 1 for runtime? No need to worry about subpackages and the like?

Also, do we need to worry about builds for musl as well?  I thought there were recent developments for dotnet on Alpine Linux which is musl based but i dont know if that can be applied here as well.

TinfoilSubmarine commented:
Regarding the subpackage versus separate templates for each component: my perspective is that, we currently have a package template that we could split out all of the components into subpackages instead of creating individual templates for each component. I'd prefer the subpackage approach because it saves setup time on the buildserver and also potential rebuilding of shared resources between the several components. Also, the tags for the `${version}-SDK` and `${version}-runtime` both point to the same commit. The tags are just a way for the devs to tag what version they are at for the runtime/SDK, not to let people download the source for runtime and SDK separately.

Has anyone tried building on musl yet? Just started a cross-build on my machine but wanted to check.

rmboggs commented:
> Regarding the subpackage versus separate templates for each component: my perspective is that, we currently have a package template that we could split out all of the components into subpackages instead of creating individual templates for each component. I'd prefer the subpackage approach because it saves setup time on the buildserver and also potential rebuilding of shared resources between the several components. Also, the tags for the `${version}-SDK` and `${version}-runtime` both point to the same commit. The tags are just a way for the devs to tag what version they are at for the runtime/SDK, not to let people download the source for runtime and SDK separately.
> 
Good points.  I didn't realize that they were pointing to the same commits.  I think it would be good to have the runtime only available so users who are using Void as a server, for example, don't have to download everything.  That being said, I don't really see much reason to break it down beyond SDK and runtime packages.  Does anyone feel differently? If not, we can try to stick with just one subpackage for this.

Anachron commented:
Why not make the SDK depend on the runtime? It would be bad to use conflicts when you can just depend on each other. Have both subpackages take the required files and use the SDK package to depend on those. So one can either only install the runtime or have the SDK with the runtime.

rmboggs commented:
> Why not make the SDK depend on the runtime? It would be bad to use conflicts when you can just depend on each other. Have both subpackages take the required files and use the SDK package to depend on those. So one can either only install the runtime or have the SDK with the runtime.

That makes sense.  If we go that route though, it may make more sense to rename the package from dotnet-sdk to simply dotnet, wouldn't it? or would it be dotnet-runtime?

Also, any thoughts on the second item?

TinfoilSubmarine commented:
>     2. The version of the SDK and runtimes are technically different (5.0.206.1 vs 5.0.9).  What would be the best/preferred approach for ensuring all packages are created with the appropriate versions?

I don't know if you can specify a different version for each subpackage, but that would be the best way I think.

TinfoilSubmarine commented:
Latest commit adds subpackages with some dependencies scraped from the Arch PKGBUILD: https://github.com/TinfoilSubmarine/void-packages/tree/dotnet-sdk

rmboggs commented:
> Latest commit adds subpackages with some dependencies scraped from the Arch PKGBUILD: https://github.com/TinfoilSubmarine/void-packages/tree/dotnet-sdk

Hey @TinfoilSubmarine ,
I see the updated structure, but it still looks like the main package is set to SDK and a new SDK subpackage.  Is this right?  Wouldn't it cause file conflicts this way?  With the dotnet executable, at least?

TinfoilSubmarine commented:
So how this is working:

`do_install()` initializes the destdir and puts all of the necessary files there.

Each `pkg_install()` then `vmove`s the files it wants into its own package. The `dotnet` package itself doesn't have any files. The `dotnet-host` package owns the executable and all of the other subpackages needing the executable pull `dotnet-host` in as a dependency.

rmboggs commented:
> So how this is working:
> 
> `do_install()` initializes the destdir and puts all of the necessary files there.
> 
> Each `pkg_install()` then `vmove`s the files it wants into its own package. The `dotnet` package itself doesn't have any files. The `dotnet-host` package owns the executable and all of the other subpackages needing the executable pull `dotnet-host` in as a dependency.

Oh, I see.  Thank you for explaining it.  I'm still trying to get used to the void package system.

I'll give it a test and see how it goes.  Thanks.

TinfoilSubmarine commented:
Probably we could make `dotnet-sdk` or `dotnet-host` the base package so that we don't have the empty meta package. Or add all of the subpackages as dependencies for the base package so that if you try to install `dotnet` it just installs all 7 subpackages.

rmboggs commented:
Actually, i was going to try to make the runtime the base package as it seems to have the basic files that every other package needs.  However, what you explained to me with the latest configuration makes sense, so maybe keep it the way you have it now.  

I'm running a build of the sdk with the new setup now to see how it goes. I'll let you know the result once it finishes but it won't be for awhile.

TinfoilSubmarine commented:
I think it may be useful for a user to be able to run `xbps-install dotnet` and just pull in everything instead of needing to install at least 3 packages to make the rest get pulled in as dependencies. [I just amended the latest commit on my fork](https://github.com/TinfoilSubmarine/void-packages/tree/dotnet-sdk) to have these changes, and also included descriptions for each subpackage.

Also @rmboggs, pro-tip about building packages with `xbps-src`: make sure to use `xbps-src pkg -C` to preserve the build environment. Especially useful when we are just messing with how we install the already-built files into the system.

I don't see much else to do here... pinging @kawaiiamber for input?

rmboggs commented:
> I think it may be useful for a user to be able to run `xbps-install dotnet` and just pull in everything instead of needing to install at least 3 packages to make the rest get pulled in as dependencies. [I just amended the latest commit on my fork](https://github.com/TinfoilSubmarine/void-packages/tree/dotnet-sdk) to have these changes, and also included descriptions for each subpackage.
> 
> Also @rmboggs, pro-tip about building packages with `xbps-src`: make sure to use `xbps-src pkg -C` to preserve the build environment. Especially useful when we are just messing with how we install the already-built files into the system.
Good to know, thank you. 
> 
> I don't see much else to do here... pinging @kawaiiamber for input?

I haven't looked at your latest changes but do we need to worry about the version differences between sdk and runtime?

TinfoilSubmarine commented:
I don't know that I fully understand the difference. Which components would be tied to which version? I saw that Arch versioned like 5.0.8.sdk205-1, e.g. something like `$runtime_version.sdk$sdk_version-$revision` for us.

rmboggs commented:
> I don't know that I fully understand the difference. Which components would be tied to which version? I saw that Arch versioned like 5.0.8.sdk205-1, e.g. something like `$runtime_version.sdk$sdk_version-$revision` for us.

That would make sense to me only if the powers that be are ok with that.

fosslinux commented:
Are you going to make a new PR so this can be better reviewed.
